doc: enhance glob pattern documentation in fs.md

mag123c opened the pull request:
The `pattern` parameter of the `fs.glob()` family was documented as a type only; nothing described which syntax the matcher accepts, what the calls return, or how matching differs across platforms. #58981 asked for exactly that.

`fsPromises.glob()` now carries the reference:

- **Pattern syntax** — wildcards (`*`, `?`, `[abc]`), globstar (`**`), brace expansion (`{a,b,c}`), and extended patterns (`+(...)`, `!(...)`)
- **Examples** — including collecting results with `Array.fromAsync`, and a table of common patterns
- **Platform considerations** — case sensitivity, path separators, the shape of returned paths, symlink traversal, and hidden files

`fs.glob()` and `fs.globSync()` cross-reference that section instead of repeating it, and the values they produce are now described.

Fixes: https://github.com/nodejs/node/issues/58981

iamstarkov reviewed (commented):
thank you, looks good to me

should it be mentioned for [path.matchesGlob(path, pattern)](https://nodejs.org/docs/latest/api/path.html#pathmatchesglobpath-pattern) as well?

mag123c commented:
@mizulu 
`path.matchesGlob()` only performs simple pattern matching, exclude functionality would need to be  implemented separately.

dmichon-msft commented on doc/api/fs.md:
```diff
@@ -1101,11 +1101,62 @@ changes:
+  console.log(entry); // .js files in src/components or src/utils
+
+// Extended glob patterns
+for await (const entry of glob('test/**/*.+(spec|test).js'))
+  console.log(entry); // Test files ending with .spec.js or .test.js
```
This will match `foo.spectest.js`, `bar.testtesttesttestspecttest.js`, etc.
Seems like a bad example of this pattern type?

jasnell commented on doc/api/fs.md:
```diff
@@ -1101,11 +1101,67 @@ changes:
+* `?` - Matches exactly one character
+* `[abc]` - Matches any character in the brackets
+* `[a-z]` - Matches any character in the range
+* `[!abc]` or `[^abc]` - Matches any character not in the brackets
```
```suggestion
* `[!abc]` or `[^abc]` - Matches any character not in the brackets
* `[!a-z]` or `[^a-z]` - Matches any character not in the range
```

mag123c replied:
Thanks for the improvement!

louwers commented on doc/api/fs.md:
````diff
@@ -1101,11 +1101,68 @@ changes:
+
+#### Examples
+
 ```mjs
````
Instead of one large code block with comments, I think it's better to split it out in multiple code snippets with documentation that is written out.

louwers commented on doc/api/fs.md:
```diff
@@ -1101,11 +1101,68 @@ changes:
+  // Note: This also matches files like 'foo.spectest.js'
+
+// More precise pattern using brace expansion
+for await (const entry of glob('test/**/*.{spec,test}.js'))
```
I would demonstrate usage of `Array.fromAsync` in one or more examples.

mag123c committed:
doc: document glob pattern syntax and behavior

The `fs.glob()` family documented its options but never described which
pattern syntax the matcher accepts, what the returned values look like,
or how matching differs across platforms.

Describe the supported wildcards, globstar, brace expansion, and
extended patterns on `fsPromises.glob()`, add common usage examples and
a platform behavior section, and point `fs.glob()` and `fs.globSync()`
at that reference. Also spell out the type and shape of the values the
callback and synchronous forms produce.

Signed-off-by: mag123c <diehreo@gmail.com>

louwers commented on doc/api/fs.md:
```diff
@@ -1101,22 +1101,171 @@ changes:
+
+**Using `Array.fromAsync` to collect results:**
+
+Convert the async iterator to an array for easier manipulation.
```
Thanks for addressing my comments!

I think the example usages can be condensed quite a bit. For example, only demonstrate how to import it once. Then you also don't have to deal with the CJS/ESM split all over the place.

What does 'for easier manipulation' mean here? Seems a bit fluff.

Maybe even just put everything in a table with only the glob pattern (inside `glob(...)` to make it more clear) and with a short description next to it.

mag123c commented:
Rebased onto main. The conflict was with the `followSymlinks` option added in #62695.

While resolving it I also corrected one bullet that had gone stale in the meantime: the "Platform Considerations" list claimed symbolic links are always followed, which is no longer accurate now that `followSymlinks` exists and defaults to `false`. It now refers to the option instead.

Could I get a re-review? cc @jasnell @lpinca @louwers

avivkeller commented on doc/api/fs.md:
```diff
@@ -1446,9 +1446,52 @@ changes:
 When `followSymlinks` is enabled, detected symbolic link cycles are not
 traversed recursively.
 
+#### Pattern Syntax
```
Can we defer to `glob(7)` or [POSIX 2.14](https://pubs.opengroup.org/onlinepubs/9799919799/utilities/V3_chap02.html#tag_19_14)?

Rather than maintain this evolving(?) syntax?

mag123c replied:
`*`, `?`, `[abc]` and the `/` + leading-dot rules are all in `glob(7)` and we match them, so I'll link out instead of listing those here. `doc/api/test.md` already does that.

The rest isn't covered by either spec, though. `**`, `{a,b}`, `+(...)`/`!(...)` and `[^abc]` aren't in POSIX 2.14 or `glob(7)`; they come from the vendored minimatch. POSIX actually says a bracket expression starting with `^` is unspecified. 2.14.3 also requires the result to be "sorted according to the collating sequence in effect in the current locale", and `fs.glob()` doesn't sort at all, so linking there would imply something we don't guarantee.

So, `glob(7)` for the basics, one small table for the extensions, and a line saying results aren't sorted. Is that okay?
